docs: Add documentation-test for README

mqasimsarfraz opened the pull request:
This change add support for testing documentation using exec-docs

blanquicet commented on .gitignore:
```diff
@@ -3,6 +3,7 @@
 *.exe~
 *.dll
 *.so
+*.log
```
Is this file generated by the `ie`? I haven't noticed it.

mqasimsarfraz replied:
Yeah, for me. I am getting ie.log once running it. Maybe it is only written if something fails?

blanquicet replied:
I don't know. Please keep track of this things so that we can provide feedback to the InnovationEngine project.

blanquicet reviewed (commented):
Looking great already!

mqasimsarfraz commented on README.md:
````diff
@@ -56,7 +56,7 @@ It is possible to download the asset for a given release and platform from the
 move the `kubectl-aks` executable to any folder in your `$PATH`.
 
 ```bash
````
Great idea! I will update the flow of the document. It is good to have uninstall step in readme and also it relates to comment [here](https://github.com/Azure/kubectl-aks/pull/51#discussion_r1376136088).

mqasimsarfraz commented on .github/workflows/kubectl-aks.yml:
```diff
@@ -258,6 +258,34 @@ jobs:
+        with:
+          go-version: '>=1.20.0' # ie requires go 1.20+
+          check-latest: true
+      - name: Get kubectl-aks from artifact
+        uses: actions/download-artifact@v3
+        with:
+          name: kubectl-aks-linux-amd64-tar-gz
+      - name: Prepare kubectl-aks binary
+        shell: bash
+        run: |
+          tar zxvf kubectl-aks-linux-amd64.tar.gz
+          chmod +x kubectl-aks
+          mv kubectl-aks /usr/local/bin
+          ls -la
```
I wanted to ensure that the binary that was built in the previous steps of the CI is actually tested since we eventually ship that in the release. But I need to ensure we clean up all other installation for it.

blanquicet replied:
Yep but currently there is no way to test the README against the just built binary. See my comment in https://github.com/Azure/kubectl-aks/pull/52#discussion_r1376196826, I think we should separate the README from others docs and the README could only be run for the releases (which makes sense as it describes how to install a released version). WDYT?

mqasimsarfraz replied:
> currently there is no way to test the README against the just built binary

I realized it is possible to pass env vars to `ie` so it can be used to add custom directory to PATH like I did [here](https://github.com/Azure/kubectl-aks/pull/51/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R110). This takes care of using the built binary by adding path of the current directory where `kubectl-aks` (built by `kubectl-aks` target) is located. In the CI, `kubectl-aks` is download at correct location and used by the tests. Shouldn't that cover all the cases?

blanquicet replied:
I don't think it covers also the README case as it doesn't make sense to run `kubectl krew install aks` (which install kubectl-aks in `~/.krew/bin/` and then run `kubectl aks version` using the binary we compiled in the build CI job.

Then, for the commands documentation, we could use such method even if I expect to have just one binary at that time given that we are cleaning up everything we install in the README. So, the only binary available in that host will be the one we downloaded from the artifacts.

mqasimsarfraz replied:
> I don't think it covers also the README case as it doesn't make sense to run kubectl krew install aks (which install kubectl-aks in ~/.krew/bin/ and then run kubectl aks version using the binary we compiled in the build CI job.

Thanks for catching this. I defined the precedence now [`krew > local > current directory`](https://github.com/Azure/kubectl-aks/pull/51/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R98) and it is working as expected [krew version](https://github.com/Azure/kubectl-aks/actions/runs/6717691754/job/18256101854#step:7:78), [specific version](https://github.com/Azure/kubectl-aks/actions/runs/6717691754/job/18256101854#step:7:108), [installed from source](https://github.com/Azure/kubectl-aks/actions/runs/6717691754/job/18256101854#step:7:159). This also ensures the `kubectl aks --help` is tested against the binary from the PR so any issues will be caught right away.

> So, the only binary available in that host will be the one we downloaded from the artifacts.

Yeah, I think for commands testing, it is much simpler. I feel like with above we should be fine for `README.md` but I am happy to test it separately. What do you say?

blanquicet replied:
I still find it too hacky. Now that you added the cleanup step for all the installation methods, I don't see the need of passing the list of paths in a specific order, there will always be only one binary installed at a time.

The only remaining problem is the `kubectl aks help`.... Mmmm I'm not sure what to do. We could either use `./kubectl-aks help` or remove it as the list of commands are already describe above in the documentation.

What do you think?

blanquicet commented on .github/workflows/kubectl-aks.yml:
```diff
@@ -268,6 +268,39 @@ jobs:
+        shell: bash
+        run: |
+          tar zxvf kubectl-aks-linux-amd64.tar.gz
+          chmod +x kubectl-aks
```
Mmmmm ... Why it doesn't have this permissions already? I don't remember doing this before.

blanquicet commented on .github/workflows/kubectl-aks.yml:
```diff
@@ -268,6 +268,39 @@ jobs:
+        run: |
+          tar zxvf kubectl-aks-linux-amd64.tar.gz
+          chmod +x kubectl-aks
+      - name: Install krew
```
How was kubectl-aks's krew installation working on the first version of this PR without krew? I thought it was already installed.

mqasimsarfraz replied:
I realized the code block only fails if the last command was unsuccessful. I opened the issue here: https://github.com/Azure/InnovationEngine/issues/125

blanquicet commented on Makefile:
```diff
@@ -69,6 +71,24 @@ install: kubectl-aks
+uninstall:
+	rm -f ~/.local/bin/kubectl-aks
+
+# Install innovation engine
```
I suggest moving this to another Makefile as it is not compiling or testing anything related to kubectl-aks.

blanquicet commented on README.md:
```diff
@@ -49,18 +49,30 @@ kubectl krew install aks
-VERSION=v0.2.0
-curl -sL https://github.com/azure/kubectl-aks/releases/latest/download/kubectl-aks-linux-amd64-${VERSION}.tar.gz | sudo tar -C /usr/local/bin -xzf - kubectl-aks
+VERSION=$(curl -s https://api.github.com/repos/azure/kubectl-aks/releases/latest | jq -r .tag_name)
+curl -sL https://github.com/azure/kubectl-aks/releases/latest/download/kubectl-aks-linux-amd64-${VERSION}.tar.gz | sudo tar -C ~/.local/bin -xzf - kubectl-aks
```
Not sure if we should use `$HOME/` or `~/`.

mqasimsarfraz commented:
> The only remaining problem is the kubectl aks help.... Mmmm I'm not sure what to do. We could either use ./kubectl-aks help or remove it as the list of commands are already describe above in the documentation.

I agree. It does look hacky. I removed testing it for now. We can reflect back later on if there is a nicer way to do it. Thanks for all the feedback. I updated the PR.

blanquicet commented on README.md:
`````diff
@@ -74,10 +86,19 @@ make install
+
+## Usage
+
+```
+kubectl aks --help
+````
+
+```
 Azure Kubernetes Service (AKS) kubectl plugin
`````
```suggestion
$ kubectl aks --help
Azure Kubernetes Service (AKS) kubectl plugin
[...]
```

blanquicet reviewed (approved):
LGTM! Thanks for working on this.

mqasimsarfraz commented:
Thanks for reviews! CI is green! I will merge this!
